### DISCUSSION TITLE
WIP-RFC: Extend Routing Capabilities

### DISCUSSION THREAD

**Discussion by CanRau** [TanStack/router#2800] [Ideas]
### Work in Progress!
> Feel free to already comment, though I'm extending this RFC with more context, examples, etc.
> Just wanted to get myself a place to get it started 🤓 

As requested by @schiller-manuel here https://github.com/TanStack/router/pull/2734#issuecomment-2480663667 I'm opening this RFC to gather everything related to possible routing features like:

- Optional Params
  - [Remix docs](https://remix.run/docs/en/main/file-conventions/routes#optional-segments)
  - As we already use `(group)` parentheses for groups we could consider `[[optional]]` double square brackets which is inspired by [NextJS](https://nextjs.org/docs/app/building-your-application/routing/dynamic-routes#optional-catch-all-segments)
- Prefixes/Suffixes & multiple params
  - draft at #2734
- Parallel Routes
- Intercepting Routes
- API Routes without the need for `apiBase`

There's also https://github.com/TanStack/router/discussions/1277 (just discovered) which I think is important to keep in mind.

Copying in this question I asked on [Discord](https://discord.com/channels/719702312431386674/1238170697650405547/1308104264480067636) (tho no one so far answered), gonna organize this whole RFC better shortly 🤓 

Hey I've got a question, what's the reason for Start to require a common apiBase prefix? In Rakkas you can name any file in the route tree to foo.api.ts or index.api.ts instead of foo.page.tsx and it's an api route. Personally I'd even prefer to be able to just respond with a raw response from any beforeLoad or loader and have it render that instead of the component which I think is kinda possible in Deno's Fresh [docs](https://fresh.deno.dev/docs/concepts/routes#-mixed-handler-and-component-route). This would allow for things like Deno is doing with their readme and imports where depending on if the Deno runtime is fetching the url or the browser they render an html page or return the raw source code. Tho personally I don't have a direct need for the latter currently, how our current project is structured where we generate Open Graph images in API routes which are located next to the page itself so kinda just append /og-image to most urls and you get the image. One thing I was thinking for this particular case is looking into using a server function and stuff its .url into the meta tag 🤔 Also our image resizing route is at /assets which might also be able to go into a server function? Or I'd have to move all those routes into /api and add redirects to not break existing links. Any tips?

**Comment by SeanCassiere** [comment]
> * API Routes without the need for apiBase

To set it up to work without the `apiBase` we'd need to move away from having a separate router (within Vinxi) for the API Routes as a whole and move over to intercepting the requests using the ssr entry and performing matching to divert the request to a matching API handler.

Short of actually doing the diversion of the intercept, an attempt was started for this - https://github.com/TanStack/router/compare/main...sean-start-api-routes-v2

~We'd also want to move away from the `export const Route = createAPIFileRoute(...)` to a different named export like `export const APIRoute = createAPIFileRoute` since the file-router in Vinxi needs to select the necessary bundle items using this key.~ (resolved with #2873)

**Comment by SeanCassiere** [comment]
> Hey I've got a question, what's the reason for Start to require a common apiBase prefix? In Rakkas you can name any file in the route tree to foo.api.ts or index.api.ts instead of foo.page.tsx and it's an api route.

This is to do with how the Vinxi router for API Routes selects the necessary routes to split out into its own bundle and not include it in the SSR bundle. At the time of implementation, Tanner wanted the SSR bundle to be separate from the API Routes bundle.

**Comment by CanRau** [comment]
I see thank you for explanation, haven't thought about vinxi's routers
